### DISCUSSION TITLE
Consider adding Qtile and version to wmname instead LG3D

### DISCUSSION THREAD

**Discussion by thingsiplay** [qtile/qtile#4149] [Show and tell]
I am not sure if this is the right category for the topic to discuss. But here we go...

The bottom part of the default script https://docs.qtile.org/en/latest/manual/config/default.html says that nobody cares about the `wmname` variable. Therefore it sets it to a specific value for irony reasons. I quote parts of it:
```
# XXX: Gasp! We're lying here. In fact, nobody really uses or cares about this
# string besides java UI toolkits;
wmname = "LG3D"
```

But that's not true. Popular programs such as *neofetch* will read and list this value. Here is what the program neofetch would output (I removed parts of the output from the command that are not interesting to us):
```
$ neofetch --backend off
DE: qtile 
WM: LG3D
```

It would be much more useful to have the current Qtile name and version in the variable instead. Here is my suggestion:
```
version = subprocess.run(['qtile', '--version'],
                         capture_output=True, text=True
                         ).stdout.strip()
wmname = f"Qtile {version}"
```

which would produce following output:
```
$ neofetch --backend off
DE: qtile 
WM: Qtile 0.22.1
```

Is there any good reason not to do this?

**Comment by thingsiplay** [comment]
A [user in Reddit have a better solution](https://www.reddit.com/r/qtile/comments/10v7kiu/quick_tip_add_qtile_version_to_wmname_to_list_it/j7gk4l7/) without using subprocess at all:

```
from libqtile.scripts.main import VERSION

wmname = f"Qtile {VERSION}"
```

**Comment by m-col** [comment]
(Just re-iterating here what I said on reddit, so that all context is on github)

The git history regarding that change was focussed largely on a small number of java apps, such as pycharm, that didn't like "Qtile" as the value. However, that was almost a decade ago so perhaps they've sorted themselves out. If you investigate the history to get the one or two main offenders and can confirm that they're working now, then +1 for changing wmname in the default config ;)

**Comment by thingsiplay** [comment]
Found this old discussion: https://github.com/qtile/qtile/issues/558

Some report that setting the name to Qtile can cause issues too. And there is another way of exporting env variable `_JAVA_AWT_WM_NONREPARENTING=1` instead. And the last comment there from 2019 report that wmname is an issue with PyCharm, but PyCharm would also run with wmname set to qtile if the mentioned env variable is set. Overall I don't use any Java app I think and don't know how it affects as of today. One of the commenter had the same idea as I had in Reddit: Just change wmname for Java applications only, with some window rulesets. And if people really still use Java programs and experience and issue, the comment could document it and tell them to switch to LG3D. 

Found another thing: https://wiki.archlinux.org/title/Bspwm#Problems_with_Java_applications
and: https://bbs.archlinux.org/viewtopic.php?id=256945

BSPWM also has same problems as Qtile. Is Qtile also a non parenting window manager? If so, maybe the _JAVA_AWT_WM_NONREPARENTING env variable is enough and even more correct. In the Archlinux discussion the guy solved the PyCharm problem with the _JAVA_AWT_WM_NONREPARENTINGenv variable.

**Comment by thingsiplay** [comment]
I have installed an inofficial PyCharm Community Edition package through Flatpak for testing, while `wmname = f"Qtile {version}"`. Application works okay, with the exception that it loses ability to focus other windows with the mouse. Keyboard focus works. I don't want go too much into details and exceptions regarding this; just saying its a little problematic. But I also tried the application with `wmname = "LG3D"` and with `import os` and `os.environ['_JAVA_AWT_WM_NONREPARENTING'] = "1"` in my config.py. PyCharm still acts the same. (_obviously relogged in to my session, so Qtile is loading correctly_)

So it seems that the program works the same, regardless of wmname or the environmental variable. I didn't try anything further than just loading window and clicking in menus and looking for something goofy to happen. It can be very well something because this is a Flatpak version. I will remove the package today, so if there is something you want me to tryout then tell me.

**Reply by m-col** [reply]
Interesting, thank you for testing it out.

So to summarise, is this correct?:

- Setting wmname to "LG3D" seems to do nothing
- any issues that "some" java apps (or older versions of apps that currently work) should be fixed by setting _JAVA_AWT_WM_NONREPARENTING alone?

**Comment by thingsiplay** [comment]
Mostly, at least from what I read in the those discussions. Problem is, this is not entirely clear, as some people report conflicting stuff. Maybe some things were changed on Python side, on Qtile or the Java environment or maybe even the specific applications such as PyCharm. And on top of it, I used the Flatpak version of an unofficial version of the application in example. There are too many variables involved to make guesses. What's interesting is, that BSPWM seem to have the same problems.

It would be good if more people report or talk about it. Would it make sense to start a poll in Reddit (or other communities too) in example? To ask who of the Qtile users use Java apps. Then maybe we could ask them to test with different wmname?

**Reply by m-col** [reply]
Yeah that might be good - even if we just hear back from 2 or 3 people to see if they can confirm whether the summary above is true or false
